lib/rack/files.rb

```diff
@@ -70,75 +71,111 @@ def serving(request, path)
-        response[0] = 200
-        range = 0..size - 1
+      if ranges.nil?
+        # No ranges:
+        ranges = [0..size - 1]
       elsif ranges.empty?
         # Unsatisfiable. Return error, and file size:
         response = fail(416, "Byte range unsatisfiable")
         response[1]["Content-Range"] = "bytes */#{size}"
         return response
-      else
-        # Partial content:
+      elsif ranges.size == 1
+        # Partial content, single range:
+        status = 206
         range = ranges[0]
-        response[0] = 206
-        response[1]["Content-Range"] = "bytes #{range.begin}-#{range.end}/#{size}"
+        headers["Content-Range"] = "bytes #{range.begin}-#{range.end}/#{size}"
         size = range.end - range.begin + 1
+      else
+        # Partial content, multiple ranges:
+        status = 206
+        headers[CONTENT_TYPE] = "multipart/byteranges; boundary=#{MULTIPART_BOUNDARY}"
+        body = Iterator.new(path, ranges, mime_type: mime_type, size: size)
+        size = body.bytesize
       end
 
-      response[2] = [response_body] unless response_body.nil?
+      headers[CONTENT_LENGTH] = size.to_s
+
+      if request.head?
+        body = []
+      else
+        body ||= Iterator.new(path, ranges, mime_type: mime_type, size: size)
+      end
 
-      response[1][CONTENT_LENGTH] = size.to_s
-      response[2] = make_body request, path, range
-      response
+      [status, headers, body]
     end
 
     class Iterator
-      attr_reader :path, :range
+      attr_reader :path, :ranges, :options
       alias :to_path :path
 
-      def initialize path, range
-        @path  = path
-        @range = range
+      def initialize(path, ranges, options)
+        @path = path
+        @ranges = ranges
+        @options = options
       end
 
       def each
         ::File.open(path, "rb") do |file|
-          file.seek(range.begin)
-          remaining_len = range.end - range.begin + 1
-          while remaining_len > 0
-            part = file.read([8192, remaining_len].min)
-            break unless part
-            remaining_len -= part.length
-
-            yield part
+          ranges.each do |range|
+            yield multipart_heading(range) if multipart?
+
+            each_range_part(file, range) do |part|
+              yield part
+            end
           end
+
+          yield "\r\n--#{MULTIPART_BOUNDARY}--\r\n" if multipart?
+        end
+      end
+
+      def bytesize
+        size = ranges.inject(0) do |sum, range|
+          sum += multipart_heading(range).bytesize if multipart?
+          sum += range.size
         end
+        size += "\r\n--#{MULTIPART_BOUNDARY}--\r\n".bytesize if multipart?
+        size
       end
 
       def close; end
-    end
 
-    private
+      private
 
-    def make_body request, path, range
-      if request.head?
-        []
-      else
-        Iterator.new path, range
+      def multipart?
+        ranges.size > 1
+      end
+
+      def multipart_heading(range)
+<<-EOF
```

Comment:
Can we encode this as a string rather than a heredoc.

Reply:
I tried to stick here with existing code styles, like
https://github.com/rack/rack/blob/93dfcdf46084760079514c858ae7391e90a7821e/lib/rack/multipart/generator.rb#L71-L89

and like examples in many test cases. 
So should I change this to string?